File: app/src/main/java/com/pleon/buyt/ui/fragment/StatesFragment.java
package com.pleon.buyt.ui.fragment;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.db.chart.animation.Animation;
import com.db.chart.model.LineSet;
import com.db.chart.view.LineChartView;
import com.pleon.buyt.R;
import com.pleon.buyt.model.Category;
import com.pleon.buyt.model.DailyCost;
import com.pleon.buyt.viewmodel.StatisticsViewModel;

import java.text.DecimalFormat;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProviders;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;

import static android.content.Intent.ACTION_TIME_TICK;
import static com.db.chart.renderer.AxisRenderer.LabelPosition.NONE;


public class StatesFragment extends Fragment {

    @BindView(R.id.chart) LineChartView lineChart;
    @BindView(R.id.chartCaption) TextView chartCaption;

    @BindView(R.id.textView3) TextView totalSpentTxvi;
    @BindView(R.id.textView) TextView averagePurchaseCostTxvi;
    @BindView(R.id.textView13) TextView mostPurchasedCatTxvi;
    @BindView(R.id.textView18) TextView numberOfPurchasesTxvi;
    @BindView(R.id.textView6) TextView maxPurchaseCostTxvi;
    @BindView(R.id.textView7) TextView minPurchaseCostTxvi;
    @BindView(R.id.textView9) TextView weekdayWithMaxPurchaseTxvi;
    @BindView(R.id.textView17) TextView storeWithMaxPurchaseTxvi;

    // Update the statistics when date changes (for example time changes from 23:59 to 00:00)
    private Date today = new Date();
    private IntentFilter timeTickIntent = new IntentFilter(ACTION_TIME_TICK);
    private BroadcastReceiver timeReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            if (new Date().getDate() != today.getDate()) {
                showStatistics();
                today = new Date();
            }
        }
    };

    private StatisticsViewModel viewModel;
    private Unbinder unbinder;

    public StatesFragment() {
        // Required empty public constructor
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        viewModel = ViewModelProviders.of(this).get(StatisticsViewModel.class);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_states, container, false);
        unbinder = ButterKnife.bind(this, view); // unbind() is required only for Fragments

        getActivity().registerReceiver(timeReceiver, timeTickIntent);

        String caption = getString(R.string.chart_caption, viewModel.getPeriod().length);
        chartCaption.setText(caption);

        showStatistics();

        return view;
    }

    private void showStatistics() {
        viewModel.getStatistics().observe(this, statistics -> {
            showGraph(statistics.getDailyCosts());

            totalSpentTxvi.setText(statistics.getTotalPurchaseCost());
            averagePurchaseCostTxvi.setText(statistics.getAveragePurchaseCost());
            if (statistics.getMostPurchasedCategoryName() != 0) {
                mostPurchasedCatTxvi.setText(statistics.getMostPurchasedCategoryName());
            }
            numberOfPurchasesTxvi.setText(statistics.getNumberOfPurchases());
            maxPurchaseCostTxvi.setText(statistics.getMaxPurchaseCost());
            minPurchaseCostTxvi.setText(statistics.getMinPurchaseCost());
            weekdayWithMaxPurchaseTxvi.setText(statistics.getWeekdayNameResWithMaxPurchases());
            storeWithMaxPurchaseTxvi.setText(statistics.getStoreNameWithMaxPurchaseCount());
        });
    }

    private void showGraph(List<DailyCost> dailyCosts) {
        lineChart.reset();

        int now = 0;
        for (DailyCost cost : dailyCosts) {
            if (cost.getCost() == -1) {
                now = cost.getDay();
                dailyCosts.remove(cost);
            }
        }

        Map<Integer, Long> dayToCostMap = new HashMap<>();
        for (DailyCost cost : dailyCosts) {
            dayToCostMap.put(cost.getDay(), cost.getCost());
        }

        LineSet dataSet = new LineSet();
        for (int i = now - viewModel.getPeriod().length + 1; i <= now; i++) {
            dataSet.addPoint("" + i, dayToCostMap.containsKey(i) ? dayToCostMap.get(i) : 0);
        }

        if (viewModel.getPeriod().length <= 20) {
            dataSet.setDotsColor(ContextCompat.getColor(getContext(), R.color.colorPrimary));
            dataSet.setDotsRadius(3);
        }
        dataSet.setSmooth(false); // TODO: Add an options in settings for the user to toggle this
        dataSet.setColor(ContextCompat.getColor(getContext(), R.color.colorPrimaryDark));
        dataSet.setThickness(2.5f);

        DecimalFormat moneyFormat = new DecimalFormat(getString(R.string.currency_format));
        lineChart.setLabelsFormat(moneyFormat);

        int[] colors = getResources().getIntArray(R.array.lineChartGradient);
        float[] steps = {0.0f, 0.5f, 1.0f};
        dataSet.setGradientFill(colors, steps);
        lineChart.addData(dataSet);
        lineChart.setXLabels(NONE);
        lineChart.show(new Animation(500));
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        getActivity().unregisterReceiver(timeReceiver);
        unbinder.unbind();
    }

    public StatisticsViewModel.Period getPeriod() {
        return viewModel.getPeriod();
    }

    public void togglePeriod() {
        viewModel.togglePeriod();
        String caption = getString(R.string.chart_caption, viewModel.getPeriod().length);
        chartCaption.setText(caption);

        showStatistics();
    }

    public Category getFilter() {
        return viewModel.getFilter();
    }

    public void setFilter(Category filter) {
        viewModel.setFilter(filter);

        showStatistics();
    }
}

File: app/src/main/kotlin/com/pleon/buyt/ui/fragment/StatesFragment.kt
package com.pleon.buyt.ui.fragment

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.Intent.ACTION_TIME_TICK
import android.content.IntentFilter
import android.os.Bundle
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.ColorRes
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import com.db.chart.animation.Animation
import com.db.chart.model.LineSet
import com.db.chart.renderer.AxisRenderer.LabelPosition.NONE
import com.pleon.buyt.R
import com.pleon.buyt.database.DailyCost
import com.pleon.buyt.database.PieSlice
import com.pleon.buyt.model.Category
import com.pleon.buyt.ui.PieChartView.Slice
import com.pleon.buyt.viewmodel.StatisticsViewModel
import kotlinx.android.synthetic.main.fragment_states.*
import java.text.DecimalFormat
import java.util.*

private const val PIE_CHART_MAX_SLICES = 5

class StatesFragment : Fragment() {

    val period get() = viewModel.period
    @ColorRes private var pieBgColor: Int = 0 // This color varies based on the app theme
    private lateinit var viewModel: StatisticsViewModel
    private val priceFormat = DecimalFormat("#,###")
    private val pieSliceColors = intArrayOf(0xffC1B435.toInt(), 0xff2DA579.toInt(),
            0xff2D71A5.toInt(), 0xffB53145.toInt(), 0xff909090.toInt())
//    private val pieSliceColors = intArrayOf(0xffC19835.toInt(), 0xffABBA33.toInt(),
//            0xff2DA579.toInt(), 0xff2D38A5.toInt(), 0xffA62D98.toInt())

    var filter: Category?
        get() = viewModel.filter
        set(filter) {
            viewModel.filter = filter
            showStatistics()
        }

    // Update the statistics when date changes (e.g. time changes from 23:59 to 00:00)
    private var today = Date()
    private val timeReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            if (Date().date != today.date) {
                showStatistics()
                today = Date()
            }
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedState: Bundle?): View {
        return inflater.inflate(R.layout.fragment_states, container, false)
    }

    override fun onViewCreated(view: View, savedState: Bundle?) {
        viewModel = ViewModelProviders.of(this).get(StatisticsViewModel::class.java)
        activity?.registerReceiver(timeReceiver, IntentFilter(ACTION_TIME_TICK))

        val typedValue = TypedValue()
        context!!.theme.resolveAttribute(R.attr.pieChartBackgroundColor, typedValue, true)
        pieBgColor = typedValue.data
        pieChart.setCell(4) // gap between slices
        pieChart.setInnerRadius(0.6f)
        pieChart.setBackGroundColor(pieBgColor)
        pieChart.setItemTextSize(21) // FIXME: text size isn't consistent across different devices
        pieChart.setAnimDuration(500)

        showStatistics()
    }

    private fun showStatistics() {
        viewModel.statistics.observe(viewLifecycleOwner, Observer { statistics ->
            showLineChart(statistics.dailyCosts!!)
            showPieChart(statistics.mostPurchasedCategories)

            textView3.text = priceFormat.format(statistics.totalPurchaseCost)
            textView.text = priceFormat.format(statistics.averagePurchaseCost)
            textView18.text = priceFormat.format(statistics.numberOfPurchases)
            textView6.text = priceFormat.format(statistics.maxPurchaseCost)
            textView7.text = priceFormat.format(statistics.minPurchaseCost)

            if (statistics.weekdayNameResWithMaxPurchases != 0) textView9.setText(statistics.weekdayNameResWithMaxPurchases)
            else textView9.text = "-"

            textView17.text = statistics.storeNameWithMaxPurchaseCount
        })
    }

    private fun showLineChart(dailyCosts: List<DailyCost>) {
        chart.reset()

        val dataSet = LineSet()
        for (dailyCost in dailyCosts) {
            dataSet.addPoint(dailyCost.date, dailyCost.totalCost.toFloat())
        }

        if (viewModel.period.length <= 20) {
            dataSet.setDotsColor(ContextCompat.getColor(context!!, R.color.colorPrimary))
            dataSet.setDotsRadius(3f)
        }
        dataSet.isSmooth = false // TODO: Add an option in settings for the user to toggle this
        dataSet.color = ContextCompat.getColor(context!!, R.color.colorPrimaryDark)
        dataSet.thickness = 2.5f

        val moneyFormat = DecimalFormat(getString(R.string.currency_format))
        chart.setLabelsFormat(moneyFormat)

        val colors = resources.getIntArray(R.array.lineChartGradient)
        val steps = floatArrayOf(0.0f, 0.5f, 1.0f)
        dataSet.setGradientFill(colors, steps)
        chart.addData(dataSet)
        chart.setXLabels(NONE)
        chart.show(Animation(500))
    }

    private fun showPieChart(pieSlices: List<PieSlice>) {
        pieChart.clearData()

        var other = 0
        for ((index, slice) in pieSlices.withIndex()) {
            if (index < PIE_CHART_MAX_SLICES - 1) {
                val sliceName = getString(Category.valueOf(slice.name).nameRes)
                pieChart.addSector(Slice(sliceName, slice.value, pieSliceColors[index]))
            } else {
                other += slice.value
            }
        }
        if (other > 0) {
            pieChart.addSector(Slice(getString(R.string.pie_chart_other), other, pieSliceColors[PIE_CHART_MAX_SLICES - 1]))
        }

        pieChart.startAnim()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        activity?.unregisterReceiver(timeReceiver)
    }

    fun togglePeriod() = viewModel.togglePeriod().also { showStatistics() }
}
